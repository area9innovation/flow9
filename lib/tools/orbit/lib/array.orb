// Array utility functions library for Orbit

// map: Apply a function to each element in an array
fn map(arr, f) = (
    if length(arr) = 0 then []
    else [f(arr[0])] + map(subrange(arr, 1, length(arr) - 1), f)
);

// filter: Keep only elements that match a predicate
fn filter(arr, pred) = (
    if length(arr) = 0 then []
    else if pred(arr[0]) then 
        [arr[0]] + filter(subrange(arr, 1, length(arr) - 1), pred)
    else 
        filter(subrange(arr, 1, length(arr) - 1), pred)
);

// fold: Accumulate a result based on each array element and an accumulator
fn fold(arr, init, f) = (
    if length(arr) = 0 then init
    else fold(subrange(arr, 1, length(arr) - 1), f(init, arr[0]), f)
);

// mapi: Map with index
fn mapi(arr, f) = (
    fn mapiHelper(arr, f, idx) = (
        if length(arr) = 0 then []
        else [f(arr[0], idx)] + mapiHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1)
    );
    mapiHelper(arr, f, 0)
);

// filteri: Filter with index
fn filteri(arr, pred) = (
    fn filteriHelper(arr, pred, idx) = (
        if length(arr) = 0 then []
        else if pred(arr[0], idx) then 
            [arr[0]] + filteriHelper(subrange(arr, 1, length(arr) - 1), pred, idx + 1)
        else 
            filteriHelper(subrange(arr, 1, length(arr) - 1), pred, idx + 1)
    );
    filteriHelper(arr, pred, 0)
);

// foldi: Fold with index
fn foldi(arr, init, f) = (
    fn foldiHelper(arr, acc, f, idx) = (
        if length(arr) = 0 then acc
        else foldiHelper(subrange(arr, 1, length(arr) - 1), f(acc, arr[0], idx), f, idx + 1)
    );
    foldiHelper(arr, init, f, 0)
);

// tail: Return all elements except the first one
fn tail(arr) = (
    if length(arr) <= 1 then []
    else subrange(arr, 1, length(arr) - 1)
);

// tailFrom: Return all elements from a given index to the end
fn tailFrom(arr, idx) = (
    if idx >= length(arr) then []
    else subrange(arr, idx, length(arr) - idx)
);

// take: Return the first n elements of an array
fn take(arr, n) = (
    if n <= 0 || length(arr) = 0 then []
    else [arr[0]] + take(subrange(arr, 1, length(arr) - 1), n - 1)
);

// iter: Call a function for each element (for side effects)
fn iter(arr, f) = (
    if length(arr) = 0 then 0 // Return 0 instead of ()
    else (
        f(arr[0]);
        iter(subrange(arr, 1, length(arr) - 1), f)
    )
);

// iteri: Iterate with index
fn iteri(arr, f) = (
    fn iteriHelper(arr, f, idx) = (
        if length(arr) = 0 then 0 // Return 0 instead of ()
        else (
            f(arr[0], idx);
            iteriHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1)
        )
    );
    iteriHelper(arr, f, 0)
);

// filtermap: Combine filter and map operations
fn filtermap(arr, f) = (
    if length(arr) = 0 then []
    else (
        let result = f(arr[0]);
        result is (
            Some(val) => [val] + filtermap(subrange(arr, 1, length(arr) - 1), f);
            None() => filtermap(subrange(arr, 1, length(arr) - 1), f);
            _ => filtermap(subrange(arr, 1, length(arr) - 1), f) // For safety
        )
    )
);

// filtermapi: Filtermap with index
fn filtermapi(arr, f) = (
    fn filtermapHelper(arr, f, idx) = (
        if length(arr) = 0 then []
        else (
            let result = f(arr[0], idx);
            result is (
                Some(val) => [val] + filtermapHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1);
                None() => filtermapHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1);
                _ => filtermapHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1) // For safety
            )
        )
    );
    filtermapHelper(arr, f, 0)
);

// removeIndex: Remove an element at a specific index
fn removeIndex(arr, idx) = (
    if idx < 0 || idx >= length(arr) then arr
    else take(arr, idx) + tailFrom(arr, idx + 1)
);

// removeRange: Remove a range of elements
fn removeRange(arr, start, count) = (
    if start < 0 || start >= length(arr) || count <= 0 then arr
    else take(arr, start) + tailFrom(arr, start + count)
);

// removeFirst: Remove the first occurrence of a specific element
fn removeFirst(arr, element) = (
    fn removeFirstHelper(before, after) = (
        if length(after) = 0 then before
        else if after[0] = element then before + subrange(after, 1, length(after) - 1)
        else removeFirstHelper(before + [after[0]], subrange(after, 1, length(after) - 1))
    );
    removeFirstHelper([], arr)
);

// removeAll: Remove all occurrences of a specific element
fn removeAll(arr, element) = (
    filter(arr, \x.x != element != 0)
);

// contains: Check if an element exists in the array
fn contains(arr, element) = (
    if length(arr) = 0 then false
    else if arr[0] = element then true
    else contains(subrange(arr, 1, length(arr) - 1), element)
);

// exists: Check if any element satisfies a predicate
fn exists(arr, pred) = (
    if length(arr) = 0 then false
    else if pred(arr[0]) then true
    else exists(subrange(arr, 1, length(arr) - 1), pred)
);

// forall: Check if all elements satisfy a predicate
fn forall(arr, pred) = (
    if length(arr) = 0 then true
    else if !pred(arr[0]) then false
    else forall(subrange(arr, 1, length(arr) - 1), pred)
);

// uniq: Remove duplicates from an array
fn uniq(arr) = (
    if length(arr) <= 1 then arr
    else (
        fn uniq_helper(remaining, result) = (
            if length(remaining) = 0 then result
            else (
                let current = remaining[0];
                let rest = tail(remaining);
                
                if contains(result, current) then
                    // Skip duplicate
                    uniq_helper(rest, result)
                else
                    // Keep this element
                    uniq_helper(rest, result + [current])
            )
        );
        uniq_helper(arr, [])
    )
);