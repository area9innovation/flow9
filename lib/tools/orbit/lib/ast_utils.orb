// AST utility functions for Orbit

// Create a properly nested AST node by folding over an array of arguments
// This function handles arrays of any length, whereas makeAst expects exactly 2 arguments
fn makeFoldedAst(op, args) = (
    let len = length(args);
    
    if len = 0 then (
        // Handle empty array case - return an appropriate default value
        // This depends on the operator, but we'll return a neutral element
        if op = "&&" then true
        else if op = "||" then false
        else args  // Return empty array as fallback
    ) else if len = 1 then (
        // For single element, just return it without creating an operation
        args[0]
    ) else if len = 2 then (
        // For exactly two elements, we can use makeAst directly
        makeAst(op, args)
    ) else (
        // For 3+ elements, fold to create a proper binary tree
        // Start with the first element and fold over the rest
        let initialAst = args[0];
        let remainingArgs = subrange(args, 1, len - 1);
        
        // Fold over the remaining arguments, combining each with the accumulator
        fold(remainingArgs, initialAst, \acc, element -> (
            makeAst(op, [acc, element])
        ))
    )
);