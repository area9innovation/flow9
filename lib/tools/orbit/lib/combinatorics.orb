// Basic combinatorial functions
fn factorial(n) = (
	if n <= 1 then 1
	else n * factorial(n - 1)
);

fn binomial_coefficient(n, k) = (
	if k < 0 || k > n then 0
	else if k == 0 || k == n then 1
	else (
		// Use multiplicative formula to avoid large intermediate values
		fn binom_helper(i, result) = (
			if i > k then result
			else binom_helper(i + 1, result * (n - k + i) / i)
		);
		binom_helper(1, 1)
	)
);

fn permutations(n, r) = (
	if r < 0 || r > n then 0
	else if r == 0 then 1
	else (
		// Calculate n! / (n-r)!
		fold(range(n-r+1, n+1), 1, \product, i.product * i)
	)
);

// Generate permutations of an array
fn generate_permutations(arr) = (
	if length(arr) <= 1 then [arr]
	else (
		let result = [];
		fn perms_helper(i, perms) = (
			if i >= length(arr) then perms
			else (
				let element = arr[i];
				let rest = remove_at_index(arr, i);
				let sub_perms = generate_permutations(rest);

				let new_perms = fold(sub_perms, perms, \acc, p.acc + [[element] + p]);
				perms_helper(i + 1, new_perms)
			)
		);
		perms_helper(0, [])
	)
);

// Generate combinations of an array
fn generate_combinations(arr, r) = (
	if r > length(arr) then []
	else if r == 0 then [[]]
	else if r == length(arr) then [arr]
	else (
		if length(arr) == 0 then []
		else (
			let first = arr[0];
			let rest = subrange(arr, 1, length(arr) - 1);

			// Combinations that include the first element
			let with_first = map(generate_combinations(rest, r - 1),
													\comb.[first] + comb);

			// Combinations that exclude the first element
			let without_first = generate_combinations(rest, r);

			with_first + without_first
		)
	)
);
