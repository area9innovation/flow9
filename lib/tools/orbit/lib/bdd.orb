import lib/rewrite;
import lib/array;
import lib/logic;

/**
 * Binary Decision Diagrams (BDDs)
 * 
 * This library implements Binary Decision Diagrams (BDDs) for Boolean formulas,
 * based on Shannon's expansion theorem. BDDs provide a canonical representation
 * for Boolean functions, which makes equivalence checking efficient.
 * 
 * BDDs are represented using the if-then-else (Ite) constructor:
 * Ite(var, then_branch, else_branch)
 * 
 * Where:
 * - var is a decision variable
 * - then_branch is the BDD when var is true
 * - else_branch is the BDD when var is false
 * - Terminal nodes are true and false
 */

// ----- VARIABLE ORDERING -----

// Get a numeric value for a variable to determine ordering
fn var_order(var) = (
    if is_not(var) then var_order(get_not_arg(var))
    else var
);

// Check if an expression is a negation
fn is_not(expr) = (
    getConstructor(expr) == "!"
);

// Get the argument of a negation
fn get_not_arg(expr) = (
    expr is (!arg => arg; _ => expr)
);

// ----- BDD CONSTRUCTION -----

// Convert a Boolean formula to a BDD
fn to_bdd(expr) = (
    expr is (
        // Terminal cases
        true => true;
        false => false;
        
        // Negation
		¬e => negate_bdd(to_bdd(e));
        
        // Conjunction (AND)
        a ∧ b => apply_and(to_bdd(a), to_bdd(b));
        
        // Disjunction (OR)
		a ∨ b => apply_or(to_bdd(a), to_bdd(b));
        
        // Already a BDD - ensure consistency
        Ite(v, t, e) => make_ite(v, to_bdd(t), to_bdd(e));
        
        // Any other expression type
        // Variable becomes a simple BDD
        v => if is_var(v) then make_ite(v, true, false) else expr;
    )
);

// Check if an expression is a variable
fn is_var(expr) = (
    let c = getConstructor(expr);
    c != "Ite" && c != "!" && c != "&&" && c != "||" && 
    expr != true && expr != false
);

// Create an Ite node with reduction
fn make_ite(var, then_case, else_case) = (
    // Rule 1: If both branches are equal, eliminate the test
    if then_case == else_case then then_case
    else Ite(var, then_case, else_case)
);

// Negate a BDD
fn negate_bdd(bdd) = (
    bdd is (
        true => false;
        false => true;
        Ite(v, t, e) => make_ite(v, negate_bdd(t), negate_bdd(e));
        _ => !bdd  // For variables or unprocessed expressions
    )
);

// Apply AND operation to two BDDs
fn apply_and(bdd1, bdd2) = (
    // Terminal cases
    if bdd1 == false || bdd2 == false then false
    else if bdd1 == true then bdd2
    else if bdd2 == true then bdd1
    else bdd1 is (
        Ite(v1, t1, e1) => (
            bdd2 is (
                Ite(v2, t2, e2) => (
                    // Choose variable to expand based on ordering
                    let v_order1 = var_order(v1);
                    let v_order2 = var_order(v2);
                    if v_order1 == v_order2 then (
                        // Same variable - combine results
                        make_ite(v1, apply_and(t1, t2), apply_and(e1, e2))
                    ) else if v_order1 < v_order2 then (
                        // v1 comes first - expand on v1
                        make_ite(v1, apply_and(t1, bdd2), apply_and(e1, bdd2))
                    ) else (
                        // v2 comes first - expand on v2
                        make_ite(v2, apply_and(bdd1, t2), apply_and(bdd1, e2))
                    )
                );
                // bdd2 is not an Ite node
                _ => make_ite(v1, apply_and(t1, bdd2), apply_and(e1, bdd2))
            )
        );
        // bdd1 is not an Ite node
        _ => bdd2 is (
            Ite(v2, t2, e2) => make_ite(v2, apply_and(bdd1, t2), apply_and(bdd1, e2));
            _ => bdd1 && bdd2  // Neither is an Ite node
        )
    )
);

// Apply OR operation to two BDDs
fn apply_or(bdd1, bdd2) = (
    // Terminal cases
    if bdd1 == true || bdd2 == true then true
    else if bdd1 == false then bdd2
    else if bdd2 == false then bdd1
    else bdd1 is (
        Ite(v1, t1, e1) => (
            bdd2 is (
                Ite(v2, t2, e2) => (
                    // Choose variable to expand based on ordering
                    let v_order1 = var_order(v1);
                    let v_order2 = var_order(v2);
                    
                    if v_order1 == v_order2 then (
                        // Same variable - combine results
                        make_ite(v1, apply_or(t1, t2), apply_or(e1, e2))
                    ) else if v_order1 < v_order2 then (
                        // v1 comes first - expand on v1
                        make_ite(v1, apply_or(t1, bdd2), apply_or(e1, bdd2))
                    ) else (
                        // v2 comes first - expand on v2
                        make_ite(v2, apply_or(bdd1, t2), apply_or(bdd1, e2))
                    )
                );
                // bdd2 is not an Ite node
                _ => make_ite(v1, apply_or(t1, bdd2), apply_or(e1, bdd2))
            )
        );
        // bdd1 is not an Ite node
        _ => bdd2 is (
            Ite(v2, t2, e2) => make_ite(v2, apply_or(bdd1, t2), apply_or(bdd1, e2));
            _ => bdd1 || bdd2  // Neither is an Ite node
        )
    )
);

// ----- BDD TO LOGIC CONVERSION -----

fn makeOr(a, b) = (
	scatter(`||`([a,b]))
);

fn makeAnd(a, b) = (
	scatter(`&&`([a,b]))
);

fn makeNot(a) = (
	makeAst("!", ([a]))
);

// Convert a BDD back to a standard Boolean formula
fn from_bdd(bdd) = (
    bdd is (
        // Terminal cases
        true => true;
        false => false;
        
        // BDD node
        Ite(v, t, e) => (
            let then_expr = from_bdd(t);
            let else_expr = from_bdd(e);
            
            // Simplify common cases
            if then_expr == true && else_expr == false then v
            else if then_expr == false && else_expr == true then makeNot(v)
            else if then_expr == true then makeOr(v, else_expr)
            else if else_expr == false then makeAnd(v, then_expr)
			else if else_expr == true then makeOr(makeNot(v), then_expr)
            else if then_expr == false then makeAnd(makeNot(v), else_expr)
            else (
				makeOr(
					makeAnd(v, then_expr),
					makeAnd(makeNot(v), else_expr)
				)
			)
        );
        
        // Variable or other expression
        _ => bdd
    )
);

// ----- VARIABLE EXTRACTION -----

// Extract all variables from a logic expression in order
fn extract_variables(expr) = (
    expr is (
        // Terminal cases
        true => [];
        false => [];
        
        // Variable
        v => if is_var(v) then [v] else [];
        
        // Negation
        !e => extract_variables(e);
        
        // AND/OR - combine variables from both sides
        a && b => uniq(sort(extract_variables(a) + extract_variables(b)));
        a || b => uniq(sort(extract_variables(a) + extract_variables(b)));
        
        // BDD node
        Ite(v, t, e) => uniq(sort([v] + extract_variables(t) + extract_variables(e)));
        
        // Fallback
        _ => []
    )
);

// ----- UTILITY FUNCTIONS -----

// Evaluate a BDD with a variable assignment
fn evaluate_bdd(bdd, assignment) = (
    bdd is (
        // Terminal cases
        true => true;
        false => false;
        
        // BDD node - follow appropriate branch based on variable value
        Ite(v, t, e) => (
            let value = assignment(v);
            if value then evaluate_bdd(t, assignment) else evaluate_bdd(e, assignment)
        );
        
        // Variable or unprocessed expression
        _ => assignment(bdd)
    )
);

// ----- EQUIVALENCE CHECKING -----

// Check if two Boolean formulas are equivalent
fn are_equivalent(expr1, expr2) = (
    let bdd1 = to_bdd(expr1);
    let bdd2 = to_bdd(expr2);
    bdd1 == bdd2
);

// ----- QUANTIFICATION OPERATIONS -----

// Variable restriction: substitute variable with a value
fn bdd_restrict(node, var, value) = (
    // Terminal cases are unchanged by restriction
    if node == true || node == false then node
    else node is (
        Ite(v, t, e) => (
            if v == var then
                // Direct substitution when we find the variable
                if value then t else e
            else (
                // For other variables, recurse on both branches
                make_ite(v, bdd_restrict(t, var, value), bdd_restrict(e, var, value))
            )
        );
        // For variables or other expressions
        _ => if node == var then value else node
    )
);

// Existential quantification: ∃v.f
fn bdd_exists(var, node) = (
    // Compute f[v→0] ∨ f[v→1]
    apply_or(
        bdd_restrict(node, var, false),
        bdd_restrict(node, var, true)
    )
);

// Universal quantification: ∀v.f
fn bdd_forall(var, node) = (
    // Compute f[v→0] ∧ f[v→1]
    apply_and(
        bdd_restrict(node, var, false),
        bdd_restrict(node, var, true)
    )
);

// Apply existential quantification for multiple variables
fn bdd_exists_multiple(vars, node) = (
    if length(vars) == 0 then node
    else bdd_exists_multiple(
        tail(vars), 
        bdd_exists(head(vars), node)
    )
);

// Apply universal quantification for multiple variables
fn bdd_forall_multiple(vars, node) = (
    if length(vars) == 0 then node
    else bdd_forall_multiple(
        tail(vars), 
        bdd_forall(head(vars), node)
    )
);

// ----- MODEL COUNTING -----

// Calculate the number of satisfying assignments for a BDD
// Simple implementation without using cache
fn bdd_count_models(node, var_count) = (
    node is (
        // No satisfying assignments for false
        false => 0;
        
        // All remaining variable assignments satisfy true
        true => (
            // 2^remaining_vars
            let result = 1;
            fn calculate_power(base, exponent, acc) = (
                if exponent <= 0 then acc
                else calculate_power(base, exponent - 1, acc * base)
            );
            calculate_power(2, var_count, 1)
        );
        
        // For BDD nodes, compute recursively
        Ite(v, t, e) => (
            // Count satisfying assignments in both branches
            let low_count = bdd_count_models(e, var_count - 1);
            let high_count = bdd_count_models(t, var_count - 1);
            
            // Sum the counts
            low_count + high_count
        );
        
        // For variables or other expressions
        _ => if node then (
            // 2^var_count
            let result = 1;
            fn calculate_power(base, exponent, acc) = (
                if exponent <= 0 then acc
                else calculate_power(base, exponent - 1, acc * base)
            );
            calculate_power(2, var_count, 1)
        ) else 0
    )
);

// Find one satisfying assignment for a BDD
fn bdd_find_satisfying_assignment(node) = (
    if node == false then
        None()  // No satisfying assignment
    else if node == true then
        Some([])  // Empty assignment works
    else node is (
        Ite(v, t, e) => (
            // Try high branch first (v=true)
            let high_result = bdd_find_satisfying_assignment(t);
            high_result is (
                Some(assignment) => Some([Pair(v, true)] + assignment);
                None() => (
                    // Try low branch (v=false)
                    let low_result = bdd_find_satisfying_assignment(e);
                    low_result is (
                        Some(assignment) => Some([Pair(v, false)] + assignment);
                        None() => None()
                    )
                )
            )
        );
        // For variables, assign true
        _ => Some([Pair(node, true)])
    )
);