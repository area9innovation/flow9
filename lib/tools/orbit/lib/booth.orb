import lib/array;

// Linear time Booth's algorithm to find canonical form in cyclic groups
fn booth_canonical(arr) = (
	if length(arr) <= 1 then arr
	else (
		let n = length(arr);

		// Create an auxiliary array which is the original array concatenated with itself
		let double_arr = arr + arr;

		// Initialize failure function array with -1
		// (We use -1 instead of conventional 0 to simplify logic)
		let f = \i. if i < 0 || i >= n then -1 else f(i);

		// Compute the min rotation index in linear time
		fn compute_min_index() = (
			// Initialize starting values
			fn find_min(i, j, k) = (
				// We're done when we've checked all rotations
				if i + k >= 2 * n then j
				else if k >= n then find_min(i + 1, j, 0)
				else (
					// Compute the current indices we're comparing
					let i_idx = (i + k) % (2 * n);
					let j_idx = (j + k) % (2 * n);

					if double_arr[i_idx] < double_arr[j_idx] then
						// i rotation is smaller, j becomes i
						find_min(i, i, k + 1)
					else if double_arr[i_idx] > double_arr[j_idx] then
						// j rotation is smaller, i moves forward
						find_min(i + k + 1, j, 0)
					else
						// Equal so far, continue comparison
						find_min(i, j, k + 1)
				)
			);

			// Start with first two positions
			find_min(0, 0, 0)
		);

		let min_start = compute_min_index();

		// Build the canonical form starting from min_start
		fn build_result(idx, result) = (
			if idx >= n then result
			else build_result(idx + 1, result + [arr[(min_start + idx) % n]])
		);

		build_result(0, [])
	)
);

