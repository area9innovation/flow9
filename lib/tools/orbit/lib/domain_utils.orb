import lib/graph;

// Compare two domains in the hierarchy
// Returns: 
//   -1 if a is more specific than b (a ⊂ ... ⊂ b)
//    0 if a and b are the same
//    1 if b is more specific than a (b ⊂ ... ⊂ a)
//    None() if a and b are not directly related
fn compareDomains(graph, a, b) = (
    if (a == b) then 0  // Same domain
    else if (isReachable(graph, a, b)) then -1  // a is more specific than b
    else if (isReachable(graph, b, a)) then 1   // b is more specific than a
    else None()  // Not directly related
);

// Find the lowest domain from a list of domains
// The lowest domain is the most specific one in the hierarchy
fn findLowestDomain(graph, domains) = (
    if (length(domains) == 0) then None()
    else if (length(domains) == 1) then Some(domains[0])
    else (
        // Start with the first domain as our current lowest
        let initialLowest = domains[0];
        
        // Compare with each other domain to find the lowest
        fold(subrange(domains, 1, length(domains) - 1), initialLowest, \currentLowest, domain -> (
            compareDomains(graph, domain, currentLowest) is (
                -1 => domain;  // This domain is more specific than current lowest
                0 => currentLowest;  // Same domain, keep current
                1 => currentLowest;  // Current lowest is more specific
                None() => (
                    // Domains are not directly related, we keep the current lowest
                    // A more sophisticated approach might look at the full hierarchy
                    currentLowest
                )
            )
        ))
    )
);

// Get the most specific domain a node belongs to
// This is the main function to be used in the OGraph system
fn getNodeMostSpecificDomain(domainGraph, og, nodeId) = (
    // Get all domains the node belongs to
    let domains = getONodeBelongsTo(og, nodeId);
    
    // Find and return the lowest domain from the list
    findLowestDomain(domainGraph, domains)
);

"Domain Hierarchy Utility Library Successfully Loaded"