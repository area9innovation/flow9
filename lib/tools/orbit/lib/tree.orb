import lib/array;
import lib/rewrite;

// Tree data structure for the Orbit system
// A binary tree used as a map with key-value pairs

// Tree node constructors:
// Empty(): represents an empty tree
// Node(key, value, left, right): a tree node with key, value, left subtree, right subtree

// Basic tree creation
fn makeTree() = (
    Empty()
);

// Tree node creation
fn makeNode(key, value, left, right) = (
    Node(key, value, left, right)
);

// Insert a key-value pair into the tree
fn insertTree(tree, key, value) = (
    tree is (
        Empty() => Node(key, value, Empty(), Empty());
        Node(k, v, left, right) => (
            if (key < k) then (
                Node(k, v, insertTree(left, key, value), right)
            ) else if (key > k) then (
                Node(k, v, left, insertTree(right, key, value))
            ) else (
                // Key already exists, update the value
                Node(k, value, left, right)
            )
        )
    )
);

// Lookup a value by key
fn lookup(tree, key) = (
    tree is (
        Empty() => None();
        Node(k, v, left, right) => (
            if (key < k) then (
                lookup(left, key)
            ) else if (key > k) then (
                lookup(right, key)
            ) else (
                Some(v)
            )
        )
    )
);

// Helper to find the minimum node in a tree
fn findMinTree(tree) = (
    tree is (
        Empty() => None();
        Node(k, v, Empty(), _) => Some(Pair(k, v));
        Node(k, v, left, _) => findMinTree(left)
    )
);

// Remove a key from the tree
fn remove(tree, key) = (
    tree is (
        Empty() => Empty();
        Node(k, v, left, right) => (
            if (key < k) then (
                Node(k, v, remove(left, key), right)
            ) else if (key > k) then (
                Node(k, v, left, remove(right, key))
            ) else (
                // Case 1: Leaf node
                if (left == Empty() && right == Empty()) then (
                    Empty()
                )
                // Case 2: Only right child
                else if (left == Empty()) then (
                    right
                )
                // Case 3: Only left child
                else if (right == Empty()) then (
                    left
                )
                // Case 4: Two children
                else (
                    // Find minimum node in right subtree
                    let minPair = findMinTree(right);
                    minPair is (
                        Some(Pair(minKey, minValue)) => (
                            // Replace current node with minimum node from right subtree
                            let newRight = remove(right, minKey);
                            Node(minKey, minValue, left, newRight)
                        );
                        None() => (
                            // This case should not occur with valid BST structure
                            // but handled for robustness
                            println("Error: findMinTree returned None on non-empty tree");
                            tree
						)
                    )
                )
            )
        )
    )
);

// Helper function to collect all entries in the tree
fn collectEntries(tree, entries) = (
    tree is (
        Empty() => entries;
        Node(k, v, left, right) => (
            // In-order traversal to maintain key ordering
            let leftEntries = collectEntries(left, entries);
            let currentEntries = arrayPush(leftEntries, Pair(k, v));
            collectEntries(right, currentEntries)
        )
    )
);

// Get all key-value pairs in the tree as an array of pairs
fn getEntries(tree) = (
    collectEntries(tree, [])
);

// Gets all keys from the tree
fn getKeys(tree) = (
    map(getEntries(tree), \pair -> pair is (Pair(first, _) => first))
);

// Gets all values from the tree
fn getValues(tree) = (
    map(getEntries(tree), \pair -> pair is (Pair(_, second) => second))
);

// Rewrite rules for tree operations
// Notice these do not reduce lookup and min from the emptry tree as an example
fn tree_rules() = (
    rules("TreeRules", [
        // Insert into empty tree
        Insert(Empty(), key, value) => Node(key, value, Empty(), Empty()),
        
        // Insert into non-empty tree - less than current key
        Insert(Node(k, v, left, right), key, value) => Node(k, v, Insert(left, key, value), right) if key < k,
        
        // Insert into non-empty tree - greater than current key
        Insert(Node(k, v, left, right), key, value) => Node(k, v, left, Insert(right, key, value)) if key > k,
        
        // Insert into non-empty tree - equal to current key (update)
        Insert(Node(k, v, left, right), key, value) => Node(k, value, left, right) if key == k,
        
        // Lookup in non-empty tree - less than current key
        Lookup(Node(k, v, left, right), key) => Lookup(left, key) if key < k,
        
        // Lookup in non-empty tree - greater than current key
        Lookup(Node(k, v, left, right), key) => Lookup(right, key) if key > k,
        
        // Lookup in non-empty tree - equal to current key
        Lookup(Node(k, v, left, right), key) => v if key == k,
        
        // Remove from empty tree
        Remove(Empty(), key) => Empty(),
        
        // Remove from non-empty tree - key less than current key
        Remove(Node(k, v, left, right), key) => Node(k, v, Remove(left, key), right) if key < k,
        
        // Remove from non-empty tree - key greater than current key
        Remove(Node(k, v, left, right), key) => Node(k, v, left, Remove(right, key)) if key > k,
        
        // Remove from non-empty tree - key matches and it's a leaf node
        Remove(Node(k, v, Empty(), Empty()), key) => Empty() if key == k,
        
        // Remove from non-empty tree - key matches and it has only left child
        Remove(Node(k, v, left, Empty()), key) => left if key == k,
        
        // Remove from non-empty tree - key matches and it has only right child
        Remove(Node(k, v, Empty(), right), key) => right if key == k,

		// TODO: Remove with two children: Just grab the left and split that
        
        // Find minimum value in a tree
        // FindMin(Empty()) => None(),
        FindMin(Node(k, v, Empty(), _)) => Pair(k, v),
        FindMin(Node(k, v, left, _)) => FindMin(left)
    ])
);
