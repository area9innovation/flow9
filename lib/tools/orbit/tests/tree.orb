import lib/tree;

// Test function to demonstrate tree operations
fn main() = (
    println("=== Binary Tree Test ===");
    
    // Create an empty tree
    let tree = makeTree();
    println("Empty tree: " + prettyOrbit(tree));
    
    // Insert some key-value pairs
    let tree1 = insertTree(tree, 5, "five");
    let tree2 = insertTree(tree1, 3, "three");
    let tree3 = insertTree(tree2, 7, "seven");
    let tree4 = insertTree(tree3, 2, "two");
    let tree5 = insertTree(tree4, 4, "four");
    
    println("\nTree after insertions: " + prettyOrbit(tree5));
    
    // Look up some values
    println("\nLookup key 3: " + prettyOrbit(lookup(tree5, 3)));
    println("Lookup key 6: " + prettyOrbit(lookup(tree5, 6)));
    
    // Get all entries
    println("\nAll entries: " + prettyOrbit(getEntries(tree5)));
    
    // Remove a node
    let tree6 = remove(tree5, 3);
    println("\nAfter removing key 3: " + prettyOrbit(tree6));
    println("Lookup key 3 after removal: " + prettyOrbit(lookup(tree6, 3)));
    
    // Test with rewrite rules
    println("\n=== Testing with rewrite rules ===");
    
    // Create a tree using rewrite rules
    let expr = makeTree();
    let expr1 = insertTree(expr, 10, "ten");
    let expr2 = insertTree(expr1, 5, "five");
    let expr3 = insertTree(expr2, 15, "fifteen");
    
    println("Original expression: " + prettyOrbit(expr3));
    
    // Apply rewrite rules
    let rules = tree_rules();
    let result = applyRules(expr3, rules);
    println("After applying rules: " + prettyOrbit(result));
    
    // Test remove with rewrite rules
    let removeExpr = remove(expr3, 5);
    let removeResult = applyRules(removeExpr, rules);
    println("\nAfter removing 5 using rules: " + prettyOrbit(removeResult));
    println("Lookup 5 after removal with rules: " + prettyOrbit(lookup(removeResult, 5)));
    
    // Test removing node with two children
    println("\n=== Testing removal of node with two children ===");
    let twoChildTree = insertTree(tree5, 6, "six");
    println("Tree with node 5 having two children: " + prettyOrbit(twoChildTree));
    let afterRemove = remove(twoChildTree, 5);
    println("After removing node 5 (has two children): " + prettyOrbit(afterRemove));
    println("All entries after removal: " + prettyOrbit(getEntries(afterRemove)));
    println("Keys in order: " + prettyOrbit(getKeys(afterRemove)));
    println("Ensuring children are preserved correctly - Lookup key 3: " + prettyOrbit(lookup(afterRemove, 3)));
    println("Looking up key 6: " + prettyOrbit(lookup(afterRemove, 6)));
    
    "Binary tree tests completed."
);

// Run the test
main();
