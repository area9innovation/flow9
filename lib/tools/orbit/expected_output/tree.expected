=== Binary Tree Test ===
Empty tree: Empty()
Tree after insertions: Node(5, "five", Node(3, "three", Node(2, "two", Empty(), Empty()), Node(4, "four", Empty(), Empty())), Node(7, "seven", Empty(), Empty()))
Lookup key 3: Some("three")
Lookup key 6: None()
All entries: [Pair(2, "two"), Pair(3, "three"), Pair(4, "four"), Pair(5, "five"), Pair(7, "seven")]
After removing key 3: Node(5, "five", Node(4, "four", Node(2, "two", Empty(), Empty()), Empty()), Node(7, "seven", Empty(), Empty()))
Lookup key 3 after removal: None()
=== Testing with rewrite rules ===
Original expression: Node(10, "ten", Node(5, "five", Empty(), Empty()), Node(15, "fifteen", Empty(), Empty()))
After applying rules: Node(10, "ten", Node(5, "five", Empty(), Empty()), Node(15, "fifteen", Empty(), Empty()))
After removing 5 using rules: Node(10, "ten", Empty(), Node(15, "fifteen", Empty(), Empty()))
Lookup 5 after removal with rules: None()
=== Testing removal of node with two children ===
Tree with node 5 having two children: Node(5, "five", Node(3, "three", Node(2, "two", Empty(), Empty()), Node(4, "four", Empty(), Empty())), Node(7, "seven", Node(6, "six", Empty(), Empty()), Empty()))
After removing node 5 (has two children): Node(6, "six", Node(3, "three", Node(2, "two", Empty(), Empty()), Node(4, "four", Empty(), Empty())), Node(7, "seven", Empty(), Empty()))
All entries after removal: [Pair(2, "two"), Pair(3, "three"), Pair(4, "four"), Pair(6, "six"), Pair(7, "seven")]
Keys in order: [2, 3, 4, 6, 7]
Ensuring children are preserved correctly - Lookup key 3: Some("three")
Looking up key 6: Some("six")
Result: "Binary tree tests completed."
